server/datastore/mysql/software_titles_test.go

```diff
@@ -0,0 +1,251 @@
+package mysql
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+	"github.com/fleetdm/fleet/v4/server/test"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSoftwareTitles(t *testing.T) {
+	ds := CreateMySQLDS(t)
+
+	cases := []struct {
+		name string
+		fn   func(t *testing.T, ds *Datastore)
+	}{
+		{"SyncHostsSoftwareTitles", testSoftwareSyncHostsSoftwareTitles},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			defer TruncateTables(t, ds)
+			c.fn(t, ds)
+		})
+	}
+}
+
+func testSoftwareSyncHostsSoftwareTitles(t *testing.T, ds *Datastore) {
+	ctx := context.Background()
+
+	cmpNameVersionCount := func(want, got []fleet.SoftwareTitle) {
+		cmp := make([]fleet.SoftwareTitle, len(got))
+		for i, sw := range got {
+			cmp[i] = fleet.SoftwareTitle{Name: sw.Name, HostsCount: sw.HostsCount}
+		}
+		require.ElementsMatch(t, want, cmp)
+	}
+
+	// this check ensures that the total number of rows in
+	// software_title_host_counts matches the expected value.
+	checkTableTotalCount := func(want int) {
+		var tableCount int
+		err := ds.writer(context.Background()).Get(&tableCount, "SELECT COUNT(*) FROM software_titles_host_counts")
+		require.NoError(t, err)
+		require.Equal(t, want, tableCount)
+	}
+
+	host1 := test.NewHost(t, ds, "host1", "", "host1key", "host1uuid", time.Now())
+	host2 := test.NewHost(t, ds, "host2", "", "host2key", "host2uuid", time.Now())
+
+	software1 := []fleet.Software{
+		{Name: "foo", Version: "0.0.1", Source: "chrome_extensions"},
+		{Name: "foo", Version: "0.0.3", Source: "chrome_extensions"},
+	}
```

Comment:
Not a blocker, but if we revisit these tests, it would be nice to include some cases that test the counts based on uniqueness of the `browser` field.